### ISSUE TITLE
Using callbacks rather than duration?

### ISSUE THREAD

**Issue by MathieuLoutre** [delucis/bgio-effects#85]
Hello! I've just started playing with this and so far this seems to respond to my needs.

That said, the duration feels like a strange API, since the whatever is happening on the client will need to match that in some ways. And if you're doing something orchestrated with multiple different duration etc. it feels like I keep changing duration between the timeout, the animation duration and effect duration.

Could providing a kind of "done" callback for the client to call when it's done be easier to manage? When the event is captured in `useEffectListener` it could come with a callback as well as the data sent via the `create` function. The timeline would wait for that callback to progress. That way, only the client has to know about the duration/visuals/sfx.

To further take away the logic from the moves, it might be practical to be able to emit events dynamically from the moves in this kind of fashion: `ctx.effects.dispatch('rollDie', { result: 6 })`

```js
const [animation, setAnimation] = useState({})

useEffectListener('rollDie', (res, done) => {
    setAnimation({
        name: 'roll', 
        res,
        done: () => {
            setAnimation({})
            done()
        }
     })
}, [])

...

{animation.name === 'roll' && <RollDie result={animation.res.result} onEnd={() => animation.done()} />}
```

And if using CSS animations with no hooks for the end, calling `done` inside a `setTimeout` should work.

I haven't had the time to look at the code properly yet so I have no idea if any of this is practical or catering for edge cases (what if "done" is never called? should there be a safeguard?) but based on the patterns I'm developing, this might be a better fit.

**Comment by MathieuLoutre** [comment]
One thing this might also enable is more control over the dispatch of the new state. `updateStateAfterEffects` is really great but I have some instances where I'd love to get the new state to arrive immediately and some delayed.

Having a callback to control the update of state would also be ideal. I have no idea how practical this is — I tried to take a look at the code but I'm not too familiar with TypeScript and I have a hard time following the structure for some reason.

**Comment by delucis** [comment]
Hi @MathieuLoutre — thanks for these ideas! It’s great to get some feedback and I’ve also been wondering about how to resolve the duplication of data that happens like needing to update the effects config every time you change the duration of an animation. I also think it would be nice if there was more control over the full state, but I’m not quite sure what API would work well for that.

I actually initially looked at having an API where your client would receive some raw stack of effects and then be able to convert that into a timeline, maintaining the client’s ownership of timings and supporting for example two clients that use different timings with the same source data. That turned out to be pretty tricky to do: it’s a much simpler API to say “add this effect at point X” than “take a stack of effects and figure out when each one should fire”.

The idea of callbacks is interesting — basically at that point the effects are just an array and each time `done()` is called the next effect is popped off the stack. However, I feel like the bidirectional dependency it creates is pretty risky. The event emitter model is designed so that if no-one is listening, that’s not a problem. Components can listen to as many or as few effects as they want. (Also, the current design supports overlapping/staggered effects, which would not be possible if each effect had to complete before the next was fired.)

A couple of ideas here, let me know what you think:

1. We could allow passing an `onEffectEnd` callback to `useEffectListener`:

    ```js
    const onEffect = res => setAnimation({ name: 'roll', res });
    const onEffectEnd = () => setAnimation({});
    useEffectListener('rollDie', onEffect, onEffectEnd, []);
    ```

    The effect listener would then call the component’s `onEffectEnd` once the duration is complete. Duration still has to be set in the effects config, but for situations where the animation is basically `setState(); setTimeout(setState, 1000);` it would effectively take care of clean-up/timing for you.

2. To help with accessing the new state early, `onEffect` (and `onEffectEnd`) could be called with the new state, e.g.:

    ```js
    useEffectListener('rollDie', (res, { G, ctx }) => {
      // use new G or ctx even if updateStateAfterEffects is true
    });
    ```

3. A `ctx.effects.dispatch('rollDie', { result: 6 })` style API could definitely be helpful. Part of the reason it is like it is currently is that it allows strict typing with TypeScript — in moves only effects that you’ve configured will be callable, in components you get typing for effect names and callbacks, and you get autocompletion etc. That helps catch errors like `ctx.effects.rollDie(6)` combined with `useEffectListener('rollDice', (res: string) => {})`. Typescript would tell you if `rollDie` or `rollDice` is correct and if the result should be a number or string.

Do any of these ideas sound helpful in your case?

**Comment by MathieuLoutre** [comment]
Thanks @delucis for the comprehensive overview. First off, I think you've assessed exactly what I was talking about and identified the main issue (what if nobody is listening, how do you remove things from the stack?).

In that light:

1. I think that's a nice little sugar and solves some of the problem and hopefully relatively straightforward to implement

2. Early state is maybe not what I'm worried about so much. Maybe my mental model is flawed but I'd like better control on the re-render, hence a more granular (per dispatch?) `updateStateAfterEffects`. For example, I have a screen where I don't want to wait for the end of the animation to update the state and one where I do. 

3.  Makes sense. I suppose it only really makes sense if we could get rid of the duration somehow.

Ultimately, `onEffectEnd` is a nice addition but I find I'm writing a lot of the same thing for each effect (and I have a few): a `setState`, a `setTimeout`, etc. and I wonder if that can be reduced. Could the state be controlled by the effect so we could "listen" directly on that? A little bit like that `{ctx.effects.rollDie && <DieRoll />}` (it's a bit of a half-baked idea because the same effect could be fired multiple times but I'm thinking out loud).

Still, I think what I'm looking for is a way to orchestrate the delivery of the state in a way I can control. For instance, it might be nice to update the state halfway through the animation.

I've started on a rough and ready effect stack implementation, rather than a timeline. 

```js
import { nanoid } from 'nanoid'

const EffectsPlugin = {
	name: 'effects',
	setup: () => {
		return {
			id: nanoid(),
			stack: []
		}
	},
	api: () => {
		const api = {
			stack: [],
			add: (name, payload) => {
				api.stack.push({ id: nanoid(), name, payload })
			}
		}

		return api
	},
	flush: ({ api }) => {
		return {
			id: nanoid(),
			stack: api.stack
		}
	}
}

export default EffectsPlugin
```

```js
import { useEffect, useState } from 'react'

function EffectsProvider ({ Board, props, listenTo }) {
	const effects = props.plugins.effects.data

	const [stack, setStack] = useState([])
	const [currentProps, setCurrentProps] = useState(props)
	const [currentStackId, setCurrentStackId] = useState(effects.id)
	const currentEffect = stack[0]

	useEffect(() => {
		// If we receive new effects, we add the ones we watch for to the stack
		if (effects.id !== currentStackId && effects.stack.length) {
			const selectedEffects = effects.stack
				.filter(({ name }) => listenTo.includes(name)) // only include effects that we listen to
				.map((e, i, list) => ({ ...e, props: i === list.length - 1 ? props : null })) // add the next props to the last element in the list

			setStack([...stack, ...selectedEffects])
			setCurrentStackId(effects.id)
		}
		// We don't want to do this if we just got a new stack
		else if (!stack.length && props !== currentProps) {
			setCurrentProps(props)
		}
	}, [props, effects, currentStackId, stack, listenTo, currentProps])

	const update = () => {
		if (currentEffect.props && currentProps !== currentEffect.props) {
			setCurrentProps(currentEffect.props)
		}
	}

	const next = (delay = 0) => {
		const nextEffect = () => {
			update()
			setStack(stack.slice(1))
		}

		if (delay > 0) {
			setStack([{ ...currentEffect, completed: true }, ...stack.slice(1)])
			setTimeout(nextEffect, delay * 1000)
		}
		else {
			nextEffect()
		}
	}

	const api = {
		...currentEffect,
		update,
		next
	}

	return <Board {...currentProps} effect={stack.length && api} />
}

export default function EffectsBoardWrapper (Board, { listenTo }) {
	return function BoardWithEffectsProvider (props) {
		return EffectsProvider({ props, Board, listenTo })
	}
}
```

There's a client-side stack of effects provided by the plugin. You provide a list of events you want to listen to and these will be added to the stack. The oldest effect in the stack is provided to the board with 2 functions:
- `update` to replace the associated props if necessary. This can be used if you want to get the new state before you run the animation or in the middle of it.
- `next` to take the effect off the stack and update the props. Optionally, you can provide a delay which will first trigger an update signalling the effect has been completed and then after the delay, take it off the stack and update the props if necessary. The delay is useful if you want to pace the end of your animation and the re-render of the board and you can watch the `completed` property to know where you are.

I'm trialling it now and though it's not perfect, it provides the kind of control I'm after. The main disadvantage is that if you say you're listening to an effect but you don't, it'll get stuck in the stack and you won't get the next props (unless you reload). The stack is also maintained through moves which is something I'm trying though I've not found a clear use for it yet.

**Comment by delucis** [comment]
@MathieuLoutre Thanks for the experimentation! I’ll post my doubts about your stack model below (although if it’s working well for you, maybe it’s just a better fit in your case, which is cool), but first some more positive counter proposals and things that seem generally a good idea.

1. We seem to agree that an `onEffectEnd` callback could help deduplicate timeouts, delegating start/end timing to the effect provider. I’d like to add this and like you say, it should be simple.

2. > “I'd like […] a more granular […] `updateStateAfterEffects`”.

    This makes a lot of sense. One option here could be something like your `update` API. That’s fairly safe: if `updateStateAfterEffects` is true, state will be delayed the whole stack, _unless_ someone calls `update` to release it early. In your example, the screen that wants state updated before the end could call `update` in response to `effects:start` or some other effect. I personally still favour receiving state early in the callbacks as this isolates things rather than globally updating across all components, but the two approaches can co-exist, so it could be left to developer choice. I would suggest exposing this via the `useEffectQueue` hook, maybe as something like `releaseGlobalState`.

3. > Could the state be controlled by the effect so we could "listen" directly on that?
    
    It should be possible to write a wrapper hook that does this already and would be even simpler with an `onEnd` callback to remove the need for specifying duration:

    ```js
    import React, { useState } from 'react';
    import { useEffectListener } from 'bgio-effects/react';

    const useEffectState = (effectName, duration, initial) => {
      const [state, setState] = useState(initial);
      const [isActive, setIsActive] = useState(false);

      useEffectListener(
        effectName,
        payload => {
          setState(payload);
          setIsActive(true);
          const timeout = setTimeout(() => setIsActive(false), duration);
          return () => clearTimeout(timeout);
        },
        []
      );

      return [state, isActive];
    };

    const Component = () => {
      const [roll, isActive] = useEffectState('rollDice', 1000, 1);

      return {isActive ? <DieRoll roll={roll} /> : null}
    };
    ```

    This could be a nice hook to add to the library, to simplify usage for many cases.

---

Ok, and here are my doubts about the stack/pop model. I’m hiding them because, like I said, if it’s working for you, that’s great, I’m just not sure about using it in this library 😃 

<details>

I’m hesitant about the reliance on `next()`, because you still end up with some tricky two-way management and potential confusion with multiple listners. Here’s an example:

- Component A and Component B both rely on `effect1`.

- Component C relies on `effect2`.

- OK, so you `listenTo: ['effect1', 'effect2']`.

So far, so good, but then

1. At some point, Component C won’t be rendered — you’re on a different screen let’s say — so you now need to make sure you update `listenTo` to remove `effect2`. To do this, you need a global context layer that allows components to register/unregister on mount/unmount and it needs to account for multiple listeners so only the *last* unregister for an effect removes it from the list.
   
    This is effectively what `useEffectListener` already provides, so it could be designed to replace the `listenTo` prop entirely, which would solve this issue and I think would be do-able.

2. Does Component A or Component B call `next()` for `effect1`?

    It should be possible to make repeated calls no-ops, but it still introduces some potentially buggy timing where multiple children can impact each other and instead of finding yourself synchronising duration between config and component, you’re synchronising between multiple components. Hopefully in a well structured app, you try to avoid this, but I can easily imagine situations where two or more deeply nested components want to respond to the same effect.

3. A one-by-one stack doesn’t support overlapping effects or would rely on the component handling the preceding event to call `next()` “early”. I am open to the idea that this level of timing detail should be left entirely to components to manage, but I do like having this flexibility currently.

</details>

**Comment by MathieuLoutre** [comment]
Hi @delucis, this is very insightful. As I've mentioned, the stack system is more rudimentary but I think it illustrates the sort of control I was after. Not really a suggestion to replace your lovely timeline. Additionally, you've pointed out some valid issues which I haven't really ran into yet, but I can see turning into problems as it scales.

Taking things out of order, on 3) that wrapper hook looks great. Seems very easy to handle and is granular while catering to your (very astute) concerns 1 & 2.

My main argument for getting rid of the duration (and therefore probably overlaps — I have no need for them in my current project and I can't quite see a use case for it right now) is that I have some effects with variable durations. For instance, if a player completes an objective, I want to display some text animation, the duration of which depends on the text, itself depending on which objective they've completed. It would be quite complex (though not impossible?) to create all the relevant effects ahead of time.

This can be solved with the `onEnd` idea but as far as I understand, it will reduce the timeline to a stack if duration is optional (if we're not aware of a duration, there can't be any effects triggered "during"). That said, I'm not sure how doable it is to take duration out of the equation within this project given the use of `rAF` to trigger the events.

On your point 2), I'm pleased that the `update` idea makes sense to you. My current approach is to almost always release the state afterwards, sometimes with a delay. The rationale is, if a player is unlocking a new thing for instance, we might want to display a "celebration" overlay and then, when the overlay is gone, update the board. For that reason, we might want to not only delay the update of the board until `onEnd` but also put a delay on it, to separate things visually. So we'd need to be able to finish an effect but release the state shortly after (at least that's the way I found it easier to model, the alternative being receiving a specific "patch" that you need to apply on `G` for the effect?).

I suppose I'm seeing effects as a way to "massage" the state updates and pace them. Specifically, what I'd love to be able to do is to trickle down state updates within a move. For instance if a move does multiple things: the player does an action (say move) and collects an object, thus completing an objective and adding a piece to the board. Ideally, I would like to play out the action (visually do the move, remove the object from the board and add to their inventory) and then play out a "celebration", and afterwards, adding the new piece to the board. I haven't found a good way to do this through boardgame.io (it's not really concerned with that at all).

I've added a way to pass new props to my makeshift stack plugin from within a move (using the likes of `current(G)`) but to be honest it feels very hacky and is quite hard to manage... It would be interesting to see if a move could result in "submoves" that we could pace through as needed, holding back and releasing state as needed by each client. Any ideas?

**Comment by delucis** [comment]
@MathieuLoutre I’ve released a new version that implements onEnd callbacks and a `useEffectState` utility hook (1 & 3 above). Still looking into providing the full new state in the callbacks and then an update method, but both of those should be fairly simple additions.

---

Something like the step-by-step “submove” updates you describe could be possible using effects. Based on your example, here’s a move (with pseudocode for the actual state management):

```js
const collectPiece = (G, ctx, piece) => {
  G.map.pieces.remove(piece);
  ctx.effects.removePiece(piece);
  G.player.inventory.add(piece);
  ctx.effects.updateInventory(G.player.inventory);
  // etc.
};
```

Then for example in an `InventoryComponent` you can listen for the new inventory state, receiving it in the effects sequence. This does duplicate some data briefly: a persistent copy in `G` and an ephemeral copy in the plugin state.

```js
const InventoryComponent = ({ G }) => {
  const [inventory, isUpdating] = useEffectState('updateInventory', G.player.inventory);
  return <ul>{inventory.map((item, idx) => <li key={idx}>{item}</li>)}</ul>;
};
```

If we make `G` etc. available in effect listener callbacks, this could also be simpler in some ways because `updateInventory` could be called with no parameters, and on the client you would just grab the latest `G` when the effect fires.

---

In related ideas (especially your desire for a bit more client control of moving between effects), I wonder if some kind of pause/resume type functionality might be helpful, so a component could request a queue to pause playback for a bit (say if a modal is open and then resume when closing it). That would allow for overlaps and would only block playback on request.

---

Regarding adding a delay after the final effect: Currently, I would suggest adding an additional “wait” effect. This could even be called in `turn.onMove` if you don’t need it to have a flexible duration, or as needed in specific moves.

```js
const config = {
  effects: {
    wait: {
      duration: 0.5,
    }
  },
};
```

```js
ctx.effects.wait();
```

No component needs to listen for the effect, but it would still delay the final state update.

That said, using the new `useEffectState` hook, I found myself wanting “gaps” between effects more often (because now an effect “end” is more meaningful, before they were just used to place the next event). To do this I do things like `ctx.effects.foo('>+0.5')` more often. I wonder if it would be helpful for an effect to have a configurable “end padding”, so a sequence of two effects with `{ duration: 3, endPad: 2 }` would be placed at 0s and 5s by default.

```
0  .  .  .  .  5  .  .  .  .  10
↑              ↑
one──────╯     two──────╯
         ╟╴pad╶╢        ╟╴pad╶╢
```

**Comment by MathieuLoutre** [comment]
Thanks @delucis, this is brilliant. I'm now using the new release and switched whatever code I had to the new hooks and making use of the end callback and I think this is working quite well now. Feels more readable and approachable.

I'm looking forward to the state release/update mechanism, particularly to deal with multiple types of clients. For instance, in the same game I have a client that will need to delay the state and another where I don't, but I still need to subscribe to the effects to know if they've finished running (for in this case, block the UI, otherwise the player can make a move and effectively reset the effect timeline).

That aside, your suggestion of using the default state and then updating it through events is really quite neat. One caveat if anybody is reading this later is that you will need to always update that state with the event afterwards so whenever you modify that part of the state, you will need to run the event. To avoid this, you can take a slightly different approach:

```js
export function useEffectStateWatcher (effectType, initialState) {
	const [state, setState] = useState(initialState)
	const [isActive, setIsActive] = useState(false)

	useEffectListener(effectType, (payload) => {
		setState(payload)
		setIsActive(true)
	}, [], () => setIsActive(false), [])

	useEffect(() => {
		setState(initialState)
	}, [initialState])

	return [state, isActive]
}
```

It's very similar to the `useEffectState` hook but with the addition that it "watches" the initial state reference and whenever it changes will set the state returned as that updated value. This means we can get updates from the effects but also from `G`. It's a little rough as is but it works if you make `initialState` a reference to `G` for instance, `G.map`.

Regarding your `pause` idea, I feel like it's a pretty good suggestions, particularly regarding your remarks about padding/spacing. This offers the possibility of "dynamic" spacing between events which I'm quite keen on. 

The padding in itself is still interesting though, particularly because a lot of visual effects will have an exit animation which will be triggered by the end of the event but will still run for some time. This makes padding very handy, but I do wonder if the timeline API as is, particularly with things like `>+0.5` can allow for a more flexible implementation than at the effect level (but you already know that I prefer having control down in the moves, as being able to specify effect duration + padding there seems more useful than at effect level and seems better than a `wait` effect itself).

Thanks again for all the work you're putting into this! Very happy to test and feedback but if you need me to start contributing to get things moving, let me know and I'm happy to make a few PRs for some of the features discussed here.

**Comment by delucis** [comment]
Good point about tracking other changes to the initial state reference. I wonder if there’s a nice API to include that (like a `watchInitialState` option). Also if full state were available in the callbacks, there could be a hook where you provide a method to run when the effect fires and on global updates, retrieving the value from the global state. Something like:

```js
const [value, isActive] = useWatchedState('rollDie', ({ G }) => G.roll);
```

> in this case, block the UI, otherwise the player can make a move and effectively reset the effect timeline

Do you have a nice way to make that seem natural for the player? Seems like delaying action, just so another player can see something, might not be the best UX. That might be solved better by having effect stacks accumulate rather than override when they arrive. That way player A could happily fire off 3 moves, and player B could still see the three moves in sequence.

The tricky thing might be making that compatible with a `pause`. I had imagined that if the dispatcher is paused, a new state update would resume it. If a modal was open and still needed to keep things paused it would subscribe to `effects:start` and call `pause` again. If effect stacks accumulated, it would be less obvious how to handle that. Auto-resuming would no longer fire `effects:start`, so that would be hard to handle, but if a component kept things paused for too long, there might be a huge sequence of effects to process when it resumed (they could always be flushed or cleared, but it might need more complex management).

> I do wonder if the timeline API as is, particularly with things like >+0.5 can allow for a more flexible implementation than at the effect level

If understand correctly, this would work similarly to the duration config — as a default value that you could override in moves. It might also be the moment to move to an options object (see https://github.com/delucis/bgio-effects/issues/41#issuecomment-692820427), so you could do `effect(payload, { endPad: 2 })` and not need a long list of parameters.

> I'm happy to make a few PRs for some of the features discussed here

Please do! #101 is also a not very exciting but useful task.

Adding an `update` method would require providing a callback in `EffectsQueueContext` similar to `clear` or `flush`, but which just calls

https://github.com/delucis/bgio-effects/blob/0a11417784aa136656439aa595962e0429e5f6a4/src/react/components.tsx#L173

Adding the general state in callbacks is also not very complex in terms of logic but requires some fiddling with the listener types, which aren’t the most friendly, so feel free to tackle that depending on how curious you are about the Typescript details.

**Comment by MathieuLoutre** [comment]
Thank you! Your `useWatchedState` hook feels a lot more elegant, getting the state in a function garanties we pull from `G` which is great. It would be great to see that implemented (as in, getting the full state in callbacks).

For the pausing/UI blocking, the UX is interesting. I've got roughly 2 types of effects:
1) the ones that are visually enhancing the experience (card animation, move animation etc.) — these usually are a result of the user's action and that the user can guess will happen from playing the move directly and benefits mostly the player who triggered the action
2) the ones that are essential for the game (dice result, new objective, new bit of story, etc.) — these are usually something triggered inside the game by a player's move or that the player couldn't know in advance and that benefits everyone playing

For 1), you don't really need to block the UI and the effects can be overwritten. The player can expect what's happening and doesn't need to be told new information. For 2) we want the players to acknowledge something that has happened internally and now affects them. In that case, the players shouldn't be able to make new moves altogether because their attention is required. In 2) the game still controls the length of the "pause" and I don't think we need user control over that (like dismissing a modal or waiting to click "ok" — this is probably better achieved by using moves and internal state).

As a result I don't think we necessarily need stacking over moves. When I tried the cumulative stack in my own naive implementation (https://github.com/delucis/bgio-effects/issues/85#issuecomment-747413816) it was quite nice to have animations being triggered over each other (I'm thinking specifically of playing a card — right now if you play a card and then another the first animation is cut short and replaced by the 2nd one, with a stack you can have them overlap) and you can flush it all as needed. However, if that makes the implementation a lot harder, it might not be worth it.

I think this is reflected in my desire to delay updates of `G` per effect rather than globally. If we imagine a UI with the user's objective and list of actions on the left and the board on the right, when 1) happens, we probably want to update quickly and while the animation is playing, the state on the left has already updated (for instance they can see they have 1 less card) and they can play the next move immediately.

When 2) happens, for instance rolling to kill a monster, we probably don't want to update the UI (actions and board) until the animation has played out because otherwise they might see whether they've failed or succeeded before it's "revealed" (and the objective might change or story might advance). In this case, it's probably in every player's interest to "watch" this play out and any action should be blocked.

I should get some time later this week to look into your suggested changes. I'm excited this is making progress already!

**Comment by delucis** [comment]
@MathieuLoutre Just spent some time adding the full state to the callbacks so you can now do:

```js
useEffectListener('effectName', (effectPayload, boardProps) => {}, []);
```

I guess the one thing a `useWatchedState` hook couldn’t do yet would be getting the initial board props before any effect has triggered (i.e. on first render). To do that, we’d have to provide board props via a React context so the hook could do something like `useBoardProps` to set an initial value.

I think for your desire to have more control over delaying `G` this should be helpful too. For the scenario you described, you could even do something like the following in the left sidebar:

```js
enhancingEffects = new Set(['playCard', 'moveItem', ...]);

const Sidebar = (props) => {
  const [localG, setLocalG] = useState(props.G);
  useEffectListener('*', (effectName, _payload, { G }) => {
    if (enhancingEffects.has(effectName) || effectName === 'effects:end') {
      setLocalG(() => G);
    }
  }, []);

  return <div>{localG.objective}</div>;
};
```

Regarding overlapping effect stacks, my main concern is that it can get pretty confusing, or at least the probability of unexpected results increases, so I’d be interested to see if with a combination of having access to full state in the callbacks, a `useWatchedState` hook, and potentially an `update` callback, they still seem necessary.

Let me know if you have any questions as you dig into the code!

**Comment by MathieuLoutre** [comment]
> I guess the one thing a `useWatchedState` hook couldn’t do yet would be getting the initial board props before any effect has triggered (i.e. on first render). To do that, we’d have to provide board props via a React context so the hook could do something like `useBoardProps` to set an initial value.

 I've just made a pull request for this, along with a draft of `useWatchedState`. A lot of questions about how to proceed in #111.

> I think for your desire to have more control over delaying `G` this should be helpful too. For the scenario you described, you could even do something like the following in the left sidebar [...] so I’d be interested to see if with a combination of having access to full state in the callbacks, a `useWatchedState` hook, and potentially an `update` callback, they still seem necessary.

Yes, It's very much in the same spirit as the watchedState idea: sharing the responsibility of a part of state between effects and moves. I think abstracting something like this into a hook, along with `useWatchedState` will definitely improve the orchestration of the release of the state to different parts of the app.

**Comment by delucis** [comment]
I think the only remaining feature from this thread is the pause/resume functionality, so I've opened #374 to track that more clearly. Feel free to correct me if there is something I've missed.
